src/Metadata/Extractor/PhpDocClassExtractor.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Metadata\Extractor;
+
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocNode;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTextNode;
+use PHPStan\PhpDocParser\Lexer\Lexer;
+use PHPStan\PhpDocParser\Parser\ConstExprParser;
+use PHPStan\PhpDocParser\Parser\PhpDocParser;
+use PHPStan\PhpDocParser\Parser\TokenIterator;
+use PHPStan\PhpDocParser\Parser\TypeParser;
+
+/**
+ * Extracts class data using a PHPDoc parser.
+ */
+final class PhpDocClassExtractor implements ClassDescriptionExtractorInterface
+{
+    private PhpDocParser $phpDocParser;
+    private Lexer $lexer;
+
+    /** @var array<string, PhpDocNode> */
+    private array $docBlocks = [];
+
+    public function __construct()
+    {
+        $this->phpDocParser = new PhpDocParser(new TypeParser(new ConstExprParser()), new ConstExprParser());
+        $this->lexer = new Lexer();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getShortDescription(string $class, array $context = []): ?string
+    {
+        if (!$docBlock = $this->getDocBlock($class)) {
+            return null;
+        }
+
+        foreach ($docBlock->children as $docChild) {
+            if ($docChild instanceof PhpDocTextNode && !empty($docChild->text)) {
+                return $docChild->text;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLongDescription(string $class, array $context = []): ?string
+    {
+        if (!$docBlock = $this->getDocBlock($class)) {
+            return null;
+        }
+
+        $summaryDiscarded = false;
+        foreach ($docBlock->children as $docChild) {
+            if ($docChild instanceof PhpDocTextNode && !empty($docChild->text)) {
+                if ($summaryDiscarded) {
+                    return $docChild->text;
+                }
+                $summaryDiscarded = true;
+            }
+        }
+
+        return null;
+    }
+
+    private function getDocBlock(string $class): ?PhpDocNode
+    {
+        if (isset($this->docBlocks[$class])) {
+            return $this->docBlocks[$class];
+        }
+
+        try {
+            $reflectionClass = new \ReflectionClass($class);
+        } catch (\ReflectionException) {
+            return null;
+        }
+
+        $rawDocNode = $reflectionClass->getDocComment();
+
+        if (!$rawDocNode) {
+            return null;
+        }
+
+        $tokens = new TokenIterator($this->lexer->tokenize($rawDocNode));
+        $phpDocNode = $this->phpDocParser->parse($tokens);
+        $tokens->consumeTokenType(Lexer::TOKEN_END);
+
+        return $this->docBlocks[$class] = $phpDocNode;
+    }
+}
```

Comment:
I do think that we can avoid this abstraction and use the phpDocParser directly in the PhpDocResourceMetadataCollectionFactory. Indeed the way I see it the metadata factories are the lowest abstraction in our system, having another abstraction adds too much code that we need to then maintain.

Reply:
I don't agree. Having this abstraction allows to reuse it elsewhere (see my other PR for GraphQL enum) and it allows to change easily the underlying implementation. Users can even write their own implementation if they have custom needs.

Reply:
my question is do we want to maintain such abstraction? Inside Symfony I'd agree but not inside API Platform.

Reply:
What is the cost of "maintaining" this class? Having this abstraction is better for a lot of reasons, I don't understand why you are so against it.

Reply:
Let me try to explain my vision on this:

ResourceMetadataFactory is and should be the single source of truth inside of API Platform when it comes to defining resources and their content. Any other abstraction should be `@internal` and not available for the end users to use. It brought problems in the past, and we're still dealing with some of these (thinking about the OperationPathResolver, or the whole filter validation feature and probably others). I'm really trying to simplify the approach for the developers and introducing a class, a new interface, an even here a new concept should really be challenged in every ways. 

Another way of seeing this is that if I'm a developper that wants to extract documentation from a class I want the developper to:
- use phpstan/doc-parser directly
- use an abstraction provided by a library 

It's not, in my opinion, API Platform's responsibility to do this and to provide such an abstraction. I might be wrong though and we probably need more opinions on this subject. This is also why we didn't create an extractor for the initial php doc implementation.

Reply:
I agree with @soyuka, let's keep this internal, it's not the responsibility of the core library to parse PHPdocs and we already provide the relevant extension points: users can just write their own metadata factories.